server/test/lib/scheduler/Schedule.spec.ts

```diff
@@ -0,0 +1,268 @@
+      schedule.assignWorkspaceToUser(USER_ID, WORKSPACE_ID_1);
+      expect(schedule.isWorkspaceCurrentlyBeingWorkedOn(WORKSPACE_ID_1)).to.equal(true);
+
+      fakeClock.tick(ONE_MINUTE / 2);
+      expect(schedule.isWorkspaceCurrentlyBeingWorkedOn(WORKSPACE_ID_1)).to.equal(true);
+
+      schedule.assignWorkspaceToUser(USER_ID, WORKSPACE_ID_2);
+      expect(schedule.isWorkspaceCurrentlyBeingWorkedOn(WORKSPACE_ID_1)).to.equal(false);
+      expect(schedule.isWorkspaceCurrentlyBeingWorkedOn(WORKSPACE_ID_2)).to.equal(true);
+
+      fakeClock.tick(ONE_MINUTE - 1);
+      expect(schedule.isWorkspaceCurrentlyBeingWorkedOn(WORKSPACE_ID_2)).to.equal(true);
+
+      fakeClock.tick(2);
+      expect(schedule.isWorkspaceCurrentlyBeingWorkedOn(WORKSPACE_ID_2)).to.equal(false);
+    });
+  });
+
+  describe("hasWorkspaceBeenWorkedOnYet", () => {
+    it("works when hasn't", () => {
+      expect(schedule.hasWorkspaceBeenWorkedOnYet(WORKSPACE_ID)).to.equal(false);
+    });
+
+    it("works when has", () => {
+      schedule.assignWorkspaceToUser(USER_ID, WORKSPACE_ID_1);
+      expect(schedule.hasWorkspaceBeenWorkedOnYet(WORKSPACE_ID_1)).to.equal(true);
+      expect(schedule.hasWorkspaceBeenWorkedOnYet(WORKSPACE_ID_2)).to.equal(false);
+    });
+  });
+
+  describe("getTimestampWorkspaceLastWorkedOn", () => {
+    it("works in straightforward case", () => {
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID, WORKSPACE_ID);
+      expect(schedule.getTimestampWorkspaceLastWorkedOn(WORKSPACE_ID)).to.equal(100);
+    });
+
+    it("works in case when assigned multiple times", () => {
+      schedule.assignWorkspaceToUser(USER_ID_1, WORKSPACE_ID);
+      fakeClock.tick(ONE_MINUTE);
+      fakeClock.tick(100);
+      schedule.assignWorkspaceToUser(USER_ID_2, WORKSPACE_ID);
+
+      expect(schedule.getTimestampWorkspaceLastWorkedOn(WORKSPACE_ID)).to.equal(ONE_MINUTE + 100);
+    });
+  });
+
+  describe("getWhichOfTheseWorkspacesWorkedOnLeastRecently", () => {
+    it("works in straightforward case", () => {
+      schedule.assignWorkspaceToUser(USER_ID_1, WORKSPACE_ID_1);
+      fakeClock.tick(ONE_MINUTE);
+      schedule.assignWorkspaceToUser(USER_ID_2, WORKSPACE_ID_2);
+
+      const result = schedule.getWhichOfTheseWorkspacesWorkedOnLeastRecently([WORKSPACE_ID_1, WORKSPACE_ID_2]);
+      expect(result).to.equal(WORKSPACE_ID_1);
+    });
+
+    it("works in case where only one has never been worked on", () => {
+      schedule.assignWorkspaceToUser(USER_ID_1, WORKSPACE_ID_1);
+
+      const result = schedule.getWhichOfTheseWorkspacesWorkedOnLeastRecently([WORKSPACE_ID_1, WORKSPACE_ID_2]);
+      expect(result).to.equal(WORKSPACE_ID_2);
+    });
+  });
+
+  describe("getTreesWorkedOnLeastRecently", () => {
+    it("works in straightforward case", done => {
+      schedule.assignWorkspaceToUser(USER_ID_1, "1-1");
+      fakeClock.tick(ONE_MINUTE);
+      schedule.assignWorkspaceToUser(USER_ID_2, "2");
```

Comment:
I suggest pulling out the Time.now() functionality in assignWorkspaceToUser, so it is an optional arg and can be passed in manually. That way this testing could be cleaner.